ocaml/testsuite/tests/typing-labeled-tuples/labeled_tuple_patterns.ml

```diff
@@ -0,0 +1,521 @@
+
+(* Full match, in order *)
+let matches =
+  let (~~(~x; ~y; ~x=x2; z)) = lt in
+  (x, y, x2, z)
+[%%expect{|
+val lt : x:int * y:int * x:int * int = (~x:1, ~y:2, ~x:3, 4)
+val matches : int * int * int * int = (1, 2, 3, 4)
+|}]
+
+(* Full match, over-bound *)
+let matches =
+  let (~~(~x; ~y; ~x; z)) = lt in
+  (x, y, z)
+[%%expect{|
+Line 2, characters 19-20:
+2 |   let (~~(~x; ~y; ~x; z)) = lt in
+                       ^
+Error: Variable x is bound several times in this matching
+|}]
+
+(* Full match, missing label *)
+let matches =
+  let (~~(~x; ~y; z)) = lt in
+  (x, y, z)
+[%%expect{|
+Line 2, characters 6-21:
+2 |   let (~~(~x; ~y; z)) = lt in
+          ^^^^^^^^^^^^^^^
+Error: This pattern matches values of type x:'a * y:'b * 'c
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+(* Full match, wrong label *)
+let matches =
+  let (~~(~x; ~y; ~w; z)) = lt in
+  (x, y, z)
+[%%expect{|
+Line 2, characters 6-25:
+2 |   let (~~(~x; ~y; ~w; z)) = lt in
+          ^^^^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type x:int * y:int * w:'a * 'b
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+(* Full match, extra label *)
+let matches =
+  let (~~(~x; ~y; ~x; ~y; z)) = lt in
+  (x, y, z)
+[%%expect{|
+Line 2, characters 6-29:
+2 |   let (~~(~x; ~y; ~x; ~y; z)) = lt in
+          ^^^^^^^^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type x:'a * y:'b * x:'c * y:'d * 'e
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+(* Full match, extra unlabeled label *)
+let matches =
+  let (~~(~x; ~y; ~x; z; w)) = lt in
+  (x, y, z)
+[%%expect{|
+Line 2, characters 6-28:
+2 |   let (~~(~x; ~y; ~x; z; w)) = lt in
+          ^^^^^^^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type x:'a * y:'b * x:'c * 'd * 'e
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+
+(* Partial match *)
+let matches =
+  let (~~(~x; ~y; .. )) = lt in
+  (x, y)
+[%%expect{|
+val matches : int * int = (1, 2)
+|}]
+
+(* Partial match, reordered *)
+let matches =
+  let (~~(~y; ~x; .. )) = lt in
+  (x, y)
+[%%expect{|
+val matches : int * int = (1, 2)
+|}]
+
+(* Partial match, reordered, over-bound *)
+let matches =
+  let (~~(~y=x; ~x; .. )) = lt in
+  x
+[%%expect{|
+Line 2, characters 13-14:
+2 |   let (~~(~y=x; ~x; .. )) = lt in
+                 ^
+Error: Variable x is bound several times in this matching
+|}]
+
+(* Partial match one *)
+let matches =
+  let (~~(~x; .. )) = lt in
+  x
+[%%expect{|
+val matches : int = 1
+|}]
+
+(* Partial match none *)
+let matches =
+  let (~~( .. )) = lt in
+  ()
+[%%expect{|
+val matches : unit = ()
+|}]
+ 
+(* Partial match all *)
+let matches =
+   let (~~( ~x; ~y; ~x=x2; z; .. )) = lt in
+   (x, y, x2, z)
+[%%expect{|
+val matches : int * int * int * int = (1, 2, 3, 4)
```

Comment:
Sorry in advance for the annoying request, but on reflection I think this should be a warning rather than an error.  It's very analogous to the warning for unused pattern variables.

See `ocaml/utils/warnings.ml` for where warnings are added.  Note that by convention we number warnings for Jane Street features down from 199 rather than up from 1.  I think I'd go with 189 for the new warning, just to avoid conflicts for some features Greta has in development that use some of the 191-197 range.

Reply:
Added this warning!